server/src/main/java/org/elasticsearch/common/settings/ConsistentSecureSettingsValidatorService.java

```diff
@@ -0,0 +1,257 @@
+        final MessageDigest digest;
+        try {
+            digest = MessageDigest.getInstance("SHA-512");
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("\"SHA-512\" algorithm is required for consistent secure settings", e);
+        }
+        if (verbatimValue instanceof InputStream) {
+            try (InputStream is = ((InputStream)verbatimValue)) {
+                try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+                    Streams.copy(is, out);
+                    digest.update(out.toByteArray());
+                }
+            } catch (IOException e) {
+                throw new RuntimeException("Exception while reading the [" + secureSetting.getKey() + "] secure setting.", e);
+            }
+        } else if (verbatimValue instanceof SecureString) {
+            try (SecureString secureString = ((SecureString)verbatimValue)) {
+                digest.update(StandardCharsets.UTF_8.encode(CharBuffer.wrap(secureString.getChars())));
+            }
+        } else {
+            throw new IllegalArgumentException("Unrecognized consistent secure setting [" + secureSetting.getKey() + "]");
+        }
+        // computing a base64 encoded unsalted SHA512 hash
+        byte[] unencodedSHA512AsByteArray = null;
+        byte[] SHA512EncodedBase64AsByteArray = null;
+        try {
+            unencodedSHA512AsByteArray = digest.digest();
+            SHA512EncodedBase64AsByteArray = Base64.getEncoder().encode(unencodedSHA512AsByteArray);
+            return CharArrays.utf8BytesToChars(SHA512EncodedBase64AsByteArray);
+        } finally {
+            if (unencodedSHA512AsByteArray != null) {
+                Arrays.fill(unencodedSHA512AsByteArray, (byte) 0);
+            }
+            if (SHA512EncodedBase64AsByteArray != null) {
+                Arrays.fill(SHA512EncodedBase64AsByteArray, (byte) 0);
+            }
+        }
+    }
+
+    private static byte[] computeSaltedHash(char[] value, byte[] salt) {
+        final int iterations = 5000;
+        final int keyLength = 512;
+        try {
+            final SecretKeyFactory skf = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA512");
+            final PBEKeySpec spec = new PBEKeySpec(value, salt, iterations, keyLength);
+            final SecretKey key = skf.generateSecret(spec);
+            return key.getEncoded();
+        } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
+            throw new RuntimeException("\"PBKDF2WithHmacSHA512\" algorithm is required for consistent secure settings", e);
+        }
+    }
+
+    private static String computePublicHash(char[] localHash, String salt) {
+        final byte[] saltedHashBytes = computeSaltedHash(localHash, salt.getBytes(StandardCharsets.UTF_8));
+        final String base64SaltedHashString = new String(Base64.getEncoder().encode(saltedHashBytes), StandardCharsets.UTF_8);
+        return salt + ":" + base64SaltedHashString;
```

Comment:
When salted hashes agree with the local hashes, the local node copies the public hashes locally so that future checks are String comparisons. Local hashes on the other hand stay put, they are never changed. From this perspective I don't think it would be beneficial to pair them together. What do you say?

Reply:
I think it is easier to understand if we have a container object and a single map. It is mainly the multiple maps that I do not care for.

Reply:
Understood, I will follow your suggestion, but because local hashes should not be published, and they are encapsulated together with the public hash, some wrangling is necessary to transpose the local map to the publishable map.